### DISCUSSION TITLE
Expo react-native and fb/google auth redirect

### DISCUSSION THREAD

**Discussion by jonathan-pyt** [supabase/supabase#2489] [Questions] [answered]
Hello ! I have been trying to get the authentication to work with Facebook and Google for several days (work perfectly in dev / published with expo), but I have a problem with the redirect with standalone app (after expo build:android)

Here is what my **Additional Redirect URLs** look like in supabase settings : 

> exp://192.168.0.21:19000,exp://exp.host/@mistes/my-app,myscheme ://*
> -----^--------------------|---------------^-------------------|-------^---------
> Work in dev (expo go)‏‏‎ ‎‏‏‎ ‎‏‏‎ ‎‏‏‎ ‎‏‏‎ ‎‏‏‎ ‎‏‏‎ ‎‏‏‎ ‎‏‏‎ ‎work if published (expo go)‏‏‎ ‎‏‏‎ ‎‏‏‎ ‎‏‏‎ ‎‏‏‎ ‎‏‏‎ ‎‏‏‎ ‎not working  after build  (myscheme is present in my app.json)
Also tested myscheme://,myscheme:///, myscheme//--/... nothing worked

At first my url looked like this :
https://auth.expo.io/@mistes974/my-app/start?authUrl=https://xxxxxxxxxx.supabase.co/auth/v1/authorize?provider=facebook&redirect_to=myscheme://&returnUrl=mysheme://

> I got this error : ValidationError: "returnUrl" must be a valid uri

Well .. i try to replace my returnUrl with : mysheme:/// (a last / ) this dismiss the error (I read about Linking / and react-navigation deep-linking) https://reactnavigation.org/docs/deep-linking/ 

> Now i got : ValidationError: "redirect_to" is not allowed 

The final url look like this : https://auth.expo.io/@mistes974/my-app/start?authUrl=https://xxxxxxxxxx.supabase.co/auth/v1/authorize?provider=facebook&returnUrl=mysheme:///
So ok ... let's remove `redirect_to` from url. and now it's working i can see the facebook popup for login. But now i never get redirect back to my app. (so auth is not complete and fail)

Anyone have a solution ?
regards,

**Comment by jonathan-pyt** [comment, marked as answer]
Finally found a solution : 

```
const socialSignIn = (provider) => {
    const returnUrl = makeRedirectUri({
      path: "/", // This were missing if you use react-navigation linking
      useProxy: false,
    });

    startAsync({
      authUrl: `${SUPABASE_URL}/auth/v1/authorize?provider=${provider}&redirect_to=${returnUrl}`,
      returnUrl: returnUrl,
    }).then(async (response) => {
      if (!response) {
        return;
      }

      const { user, session, error } = await auth.signIn({
        refreshToken: response.params?.refresh_token,
      });

      console.log(user, session, error);

    });
  };

```

```
import * as Linking from 'expo-linking';
const prefix = Linking.createURL('/');

function App() {
  const linking = {
    prefixes: [prefix],
  };

  return (
    <NavigationContainer linking={linking}>
      {/* content */}
    </NavigationContainer>
  );
}
```

**Reply by cloudorbush** [reply]
How does one create the redirect when signing in without a provider?

**Reply by Seejey01** [reply]
For newer `supabase` client, instead of `auth.signIn` you'll make use of `auth.setSession` like below 👇
```
supabase.auth.setSession({
  refresh_token: response.params?.refresh_token,
  access_token: response.params?.access_token,
})
```

**Comment by Johnrobmiller** [comment]
Auth seems like the sort of thing that is supposed to work out of the box -- with that said, this seems like a lot of setup and debugging.

For those who don't want to go through the process of configuring React Navigation to work with Supabase, you can probably just use firebase or Clerk for auth, and then continue using Supabase for the db. However, if you need row-level security, and for React Native users you probably do, then this plan might have more challenges
